File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # approach: binary search while checking for new min
        
        result = float('inf')
        l = 0
        r = len(nums) - 1
        
        while l <= r:
            m = (l + r) // 2
            
            result = min(result, nums[m])
            
            if nums[m] > nums[r]:
                l = m + 1
            else:
                r = m - 1
                
        return result

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        int l = 0;
        int r = nums.length - 1;
        int min = nums[0];
        
        while (l <= r) {
            int m = (l + r) / 2;
            
            min = Math.min(min, nums[m]);
            
            if (nums[m] > nums[r]) {
                l = m + 1;
            } else if (nums[m] < nums[r]) {
                r = m - 1;
            } else {
                return min;
            }
        }
        
        return min;
    }
}